src/fundus/parser/utility.py

```diff
@@ -154,25 +155,34 @@ def extract_nodes(selector: XPath, node_type: Type[Node]) -> List[Node]:
 
 _ld_node_selector = XPath("//script[@type='application/ld+json']")
 _json_pattern = re.compile(r"(?P<json>{[\s\S]*}|\[\s*{[\s\S]*}\s*](?!\s*}))")
+_json_undefined = re.compile(r'(?P<key>"[^"]*?"):\s*undefined')
 
 
 def sanitize_json(text: str) -> Optional[str]:
     # capture only content enclosed as follows: {...} or [{...}]
     match = re.search(_json_pattern, text)
-    if match is not None and (sanitized := match.group("json")):
-        return sanitized
-    return None
-
-
-def extract_json_from_dom(root: lxml.html.HtmlElement, selector: XPath) -> Iterable[Dict[str, Any]]:
-    json_nodes = selector(root)
-    jsons = []
-    for node in json_nodes:
-        json_content = sanitize_json(node.text_content()) or ""
-        try:
-            jsons.append(json.loads(json_content))
-        except json.JSONDecodeError as error:
-            logger.debug(f"Encountered {error!r} during JSON parsing")
+    if match is None or not (sanitized := match.group("json")):
+        return None
+
+    # substitute "bad" values
+    sanitized = re.sub(_json_undefined, r"\g<key>:null", sanitized)
```

Comment:
This is possible? 🤯 You don't want to know how often I worked around that to avoid selecting text I don't want to loose due to substitution...

Reply:
I feel you 😄 regex is a rabbit hole and every time I have to use it I find myself half an hour and 30 open tabs later scratching my head.